chemclipse/plugins/org.eclipse.chemclipse.msd.model/src/org/eclipse/chemclipse/msd/model/matrix/ExtractedMatrix.java

```diff
@@ -0,0 +1,189 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Lablicate GmbH.
+ * 
+ * All rights reserved.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v1.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors:
+ * Lorenz Gerber - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.chemclipse.msd.model.matrix;
+
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.chemclipse.msd.model.core.IIon;
+import org.eclipse.chemclipse.msd.model.core.IIonBounds;
+import org.eclipse.chemclipse.msd.model.core.IScanMSD;
+import org.eclipse.chemclipse.msd.model.core.selection.IChromatogramSelectionMSD;
+import org.eclipse.chemclipse.msd.model.implementation.Ion;
+
+public class ExtractedMatrix implements IExtractedMatrix {
+	
+	private IChromatogramSelectionMSD selection;
+	private List<IScanMSD> scans;
+	private int numberOfScans;
+	private int numberOfIons;
+	private int startIon;
+	private int stopIon;
+	private float[] signal;
+	private boolean highRes = false;
+	
+	public ExtractedMatrix(IChromatogramSelectionMSD chromatogramSelection) {
+		this.selection = chromatogramSelection;
+		this.numberOfScans = selection.getStopScan() - selection.getStartScan() - 1;
+		this.scans = extractScans();
+		if(checkHighRes( 10)) {
+			highRes = true;
+			System.out.println("This is highres MS Data and should be binned first");
+		} else {
+			this.startIon = getMinMz();
+			this.stopIon = getMaxMz();
+			this.numberOfIons = this.stopIon - this.startIon + 1;
+			signal = new float[numberOfScans * (this.stopIon - this.startIon + 1)];
+			try {
+				for(int i = 0; i < numberOfScans; i++) {
+					for(int j = 0; j < scans.get(i).getIons().size(); j++) {
```

Comment:
it would be better to extract

1. the scan (it is required in the inner loop anyways)
2. the size so it must not be retrieved on each iteration

beside that j could be renamed to 'ionIndex'